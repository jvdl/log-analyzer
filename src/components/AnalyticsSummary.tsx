import { Analytics } from '../types/log';

const TOP_COUNT = 3;

export function AnalyticsSummary ({ analytics }: { analytics: Analytics | null}) {
  if (!analytics) {
    return null;
  }

  const topRemoteAddr = analytics.countByRemoteAddr.slice(0, TOP_COUNT);
  const topPaths = analytics.countByPath.slice(0, TOP_COUNT);
  const uniqueRemoteAddrCount = analytics.countByRemoteAddr.length;

  return (<>
    <h2>Analytics summary</h2>
      <ul>
        <li>Number of unique IP addresses: <strong>{uniqueRemoteAddrCount}</strong></li>
        <li>Top {TOP_COUNT} IP addresses:
          <ol>
            {topRemoteAddr.map(({ remote_addr, count }) => (
              <li key={remote_addr}><strong>{remote_addr}</strong> <em>({count} requests)</em></li>
            ))}
          </ol>
        </li>
        <li>Top {TOP_COUNT} most requested paths:
          <ol>
            {topPaths.map(({ path, count }) => (
              <li key={path}><strong>{path}</strong> <em>({count} requests)</em></li>
            ))}
          </ol>
        </li>
      </ul>
  </>);
}
