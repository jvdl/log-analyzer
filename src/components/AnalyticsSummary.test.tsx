import { AnalyticsSummary } from './AnalyticsSummary';
import renderer from 'react-test-renderer';
import { describe, it, expect } from 'vitest';


describe(AnalyticsSummary, () => {
  const defaultAnalytics = {
    groupedByRemoteAddr: {
      '123.456.789.0': [],
      '123.456.789.1': [],
      '123.456.789.2': [],
      '123.456.789.3': [],
      '123.456.789.4': [],
    },
    groupedByPath: {},
    countByRemoteAddr: [
      { remote_addr: '123.456.789.0', count: 5 },
      { remote_addr: '123.456.789.1', count: 4 },
      { remote_addr: '123.456.789.2', count: 3 },
      { remote_addr: '123.456.789.3', count: 2 },
      { remote_addr: '123.456.789.4', count: 1 },
    ],
    countByPath: [
      { path: '/path/1', count: 5 },
      { path: '/path/2', count: 4 },
      { path: '/path/3', count: 3 },
      { path: '/path/4', count: 2 },
      { path: '/path/5', count: 1 },
    ],
  };
  it('renders correctly with expected data', () => {
    const tree = renderer
      .create(<AnalyticsSummary analytics={defaultAnalytics} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
  });

  it('renders correctly with null data', () => {
    const tree = renderer
      .create(<AnalyticsSummary analytics={null} />)
      .toJSON();
      expect(tree).toBeNull();
  });

  it('renders correctly with less than the required number of results', () => {
    const analytics = {
      ...defaultAnalytics,
      countByRemoteAddr: defaultAnalytics.countByRemoteAddr.slice(0, 2),
      countByPath: defaultAnalytics.countByPath.slice(0, 2),
    };
    const tree = renderer
      .create(<AnalyticsSummary analytics={analytics} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
  });
});
