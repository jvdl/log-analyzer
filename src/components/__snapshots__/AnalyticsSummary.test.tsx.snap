// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AnalyticsSummary > renders correctly with expected data 1`] = `
[
  <h2>
    Analytics summary
  </h2>,
  <ul>
    <li>
      Number of unique IP addresses: 
      <strong>
        5
      </strong>
    </li>
    <li>
      Top 
      3
       IP addresses:
      <ol>
        <li>
          <strong>
            123.456.789.0
          </strong>
           
          <em>
            (
            5
             requests)
          </em>
        </li>
        <li>
          <strong>
            123.456.789.1
          </strong>
           
          <em>
            (
            4
             requests)
          </em>
        </li>
        <li>
          <strong>
            123.456.789.2
          </strong>
           
          <em>
            (
            3
             requests)
          </em>
        </li>
      </ol>
    </li>
    <li>
      Top 
      3
       most requested paths:
      <ol>
        <li>
          <strong>
            /path/1
          </strong>
           
          <em>
            (
            5
             requests)
          </em>
        </li>
        <li>
          <strong>
            /path/2
          </strong>
           
          <em>
            (
            4
             requests)
          </em>
        </li>
        <li>
          <strong>
            /path/3
          </strong>
           
          <em>
            (
            3
             requests)
          </em>
        </li>
      </ol>
    </li>
  </ul>,
]
`;

exports[`AnalyticsSummary > renders correctly with less than the required number of results 1`] = `
[
  <h2>
    Analytics summary
  </h2>,
  <ul>
    <li>
      Number of unique IP addresses: 
      <strong>
        2
      </strong>
    </li>
    <li>
      Top 
      3
       IP addresses:
      <ol>
        <li>
          <strong>
            123.456.789.0
          </strong>
           
          <em>
            (
            5
             requests)
          </em>
        </li>
        <li>
          <strong>
            123.456.789.1
          </strong>
           
          <em>
            (
            4
             requests)
          </em>
        </li>
      </ol>
    </li>
    <li>
      Top 
      3
       most requested paths:
      <ol>
        <li>
          <strong>
            /path/1
          </strong>
           
          <em>
            (
            5
             requests)
          </em>
        </li>
        <li>
          <strong>
            /path/2
          </strong>
           
          <em>
            (
            4
             requests)
          </em>
        </li>
      </ol>
    </li>
  </ul>,
]
`;
